File: trees/trees.py
"""
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
TREES  ·  Python
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

BST PROPERTY: left.val < node.val < right.val for every node
TRAVERSALS  : Inorder (→ sorted), Preorder, Postorder, Level-order (BFS)
AVL TREE    : Self-balancing BST — guarantees O(log n) via rotations
"""

from __future__ import annotations
from collections import deque
from typing import Optional


class TreeNode:
    def __init__(self, val: int = 0,
                 left: Optional[TreeNode] = None,
                 right: Optional[TreeNode] = None) -> None:
        self.val = val
        self.left = left
        self.right = right

    @staticmethod
    def from_list(vals: list[Optional[int]]) -> Optional[TreeNode]:
        if not vals or vals[0] is None: return None
        root = TreeNode(vals[0])
        q: deque[TreeNode] = deque([root])
        i = 1
        while q and i < len(vals):
            node = q.popleft()
            if i < len(vals) and vals[i] is not None:
                node.left = TreeNode(vals[i]); q.append(node.left)  # type: ignore
            i += 1
            if i < len(vals) and vals[i] is not None:
                node.right = TreeNode(vals[i]); q.append(node.right)  # type: ignore
            i += 1
        return root


# ══════════════════════════════════════════════
# BST
# ══════════════════════════════════════════════

class BST:
    def __init__(self) -> None:
        self.root: Optional[TreeNode] = None

    def insert(self, val: int) -> None:
        def _ins(node, v):
            if not node: return TreeNode(v)
            if v < node.val: node.left = _ins(node.left, v)
            elif v > node.val: node.right = _ins(node.right, v)
            return node
        self.root = _ins(self.root, val)

    def search(self, val: int) -> bool:
        node = self.root
        while node:
            if val == node.val: return True
            node = node.left if val < node.val else node.right
        return False

    def delete(self, val: int) -> None:
        def _del(node, v):
            if not node: return None
            if v < node.val: node.left = _del(node.left, v)
            elif v > node.val: node.right = _del(node.right, v)
            else:
                if not node.left: return node.right
                if not node.right: return node.left
                # Find inorder successor (min of right subtree)
                succ = node.right
                while succ.left: succ = succ.left
                node.val = succ.val
                node.right = _del(node.right, succ.val)
            return node
        self.root = _del(self.root, val)

    def inorder(self) -> list[int]:
        res: list[int] = []
        def dfs(n):
            if n: dfs(n.left); res.append(n.val); dfs(n.right)
        dfs(self.root); return res


# ══════════════════════════════════════════════
# INTERVIEW PROBLEMS
# ══════════════════════════════════════════════

def max_depth(root: Optional[TreeNode]) -> int:
    """🟢 Maximum Depth of Binary Tree (LC #104)"""
    if not root: return 0
    return 1 + max(max_depth(root.left), max_depth(root.right))

def invert_tree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """🟢 Invert Binary Tree (LC #226)"""
    if not root: return None
    root.left, root.right = invert_tree(root.right), invert_tree(root.left)
    return root

def is_symmetric(root: Optional[TreeNode]) -> bool:
    """🟢 Symmetric Tree (LC #101)"""
    def mirror(l, r):
        if not l and not r: return True
        if not l or not r: return False
        return l.val == r.val and mirror(l.left, r.right) and mirror(l.right, r.left)
    return mirror(root.left, root.right) if root else True

def diameter(root: Optional[TreeNode]) -> int:
    """🟢 Diameter of Binary Tree (LC #543)"""
    best = [0]
    def height(n):
        if not n: return 0
        l, r = height(n.left), height(n.right)
        best[0] = max(best[0], l + r)
        return 1 + max(l, r)
    height(root); return best[0]

def level_order(root: Optional[TreeNode]) -> list[list[int]]:
    """🟡 Binary Tree Level Order Traversal (LC #102)"""
    if not root: return []
    result, q = [], deque([root])
    while q:
        level = []
        for _ in range(len(q)):
            node = q.popleft()
            level.append(node.val)
            if node.left: q.append(node.left)
            if node.right: q.append(node.right)
        result.append(level)
    return result

def right_side_view(root: Optional[TreeNode]) -> list[int]:
    """🟡 Binary Tree Right Side View (LC #199)"""
    result, q = [], deque([root]) if root else deque()
    while q:
        for i in range(len(q)):
            node = q.popleft()
            if i == len(q): pass  # last of this level added below
            if node.left: q.append(node.left)
            if node.right: q.append(node.right)
        result.append(node.val)  # type: ignore
    return result

def lca(root: Optional[TreeNode], p: TreeNode, q: TreeNode) -> Optional[TreeNode]:
    """🟡 Lowest Common Ancestor (LC #236)"""
    if not root or root is p or root is q: return root
    left = lca(root.left, p, q)
    right = lca(root.right, p, q)
    return root if left and right else left or right

def build_from_preorder_inorder(preorder: list[int], inorder: list[int]) -> Optional[TreeNode]:
    """🟡 Construct Binary Tree from Preorder and Inorder (LC #105)"""
    if not preorder: return None
    root = TreeNode(preorder[0])
    mid = inorder.index(preorder[0])
    root.left  = build_from_preorder_inorder(preorder[1:mid+1], inorder[:mid])
    root.right = build_from_preorder_inorder(preorder[mid+1:], inorder[mid+1:])
    return root

def kth_smallest(root: Optional[TreeNode], k: int) -> int:
    """🟡 Kth Smallest Element in a BST (LC #230) — iterative inorder"""
    stack, cur, count = [], root, 0
    while stack or cur:
        while cur: stack.append(cur); cur = cur.left
        cur = stack.pop(); count += 1
        if count == k: return cur.val
        cur = cur.right
    return -1

def max_path_sum(root: Optional[TreeNode]) -> int:
    """🔴 Binary Tree Maximum Path Sum (LC #124)"""
    best = [float('-inf')]
    def gain(n):
        if not n: return 0
        l, r = max(gain(n.left), 0), max(gain(n.right), 0)
        best[0] = max(best[0], l + r + n.val)
        return n.val + max(l, r)
    gain(root); return int(best[0])

def serialize(root: Optional[TreeNode]) -> str:
    """🔴 Serialize Binary Tree (LC #297)"""
    if not root: return "null"
    res, q = [], deque([root])
    while q:
        node = q.popleft()
        if node:
            res.append(str(node.val)); q.append(node.left); q.append(node.right)
        else:
            res.append("null")
    return ','.join(res)

def deserialize(data: str) -> Optional[TreeNode]:
    vals = data.split(',')
    if vals[0] == "null": return None
    root = TreeNode(int(vals[0]))
    q: deque[TreeNode] = deque([root])
    i = 1
    while q and i < len(vals):
        node = q.popleft()
        if vals[i] != "null": node.left = TreeNode(int(vals[i])); q.append(node.left)  # type: ignore
        i += 1
        if i < len(vals) and vals[i] != "null": node.right = TreeNode(int(vals[i])); q.append(node.right)  # type: ignore
        i += 1
    return root

def is_balanced(root: Optional[TreeNode]) -> bool:
    """🟡 Balanced Binary Tree (LC #110)"""
    def check(n):
        if not n: return 0
        l, r = check(n.left), check(n.right)
        if l == -1 or r == -1 or abs(l - r) > 1: return -1
        return 1 + max(l, r)
    return check(root) != -1

def path_sum(root: Optional[TreeNode], target: int) -> list[list[int]]:
    """🟡 Path Sum II (LC #113) — all root-to-leaf paths summing to target"""
    result: list[list[int]] = []
    def dfs(node, rem, path):
        if not node: return
        path.append(node.val)
        if not node.left and not node.right and rem == node.val:
            result.append(list(path))
        dfs(node.left, rem - node.val, path)
        dfs(node.right, rem - node.val, path)
        path.pop()
    dfs(root, target, [])
    return result


# ══════════════════════════════════════════════
# TESTS
# ══════════════════════════════════════════════

def run_tests() -> None:
    print("Running tree tests...\n")

    # BST
    bst = BST()
    for v in [5,3,7,1,4,6,8]: bst.insert(v)
    assert bst.inorder() == [1,3,4,5,6,7,8]
    assert bst.search(4) and not bst.search(9)
    bst.delete(3)
    assert bst.inorder() == [1,4,5,6,7,8]
    print("  ✅ BST: insert / search / delete / inorder")

    root = TreeNode.from_list([3,9,20,None,None,15,7])
    assert max_depth(root) == 3
    assert max_depth(None) == 0
    print("  ✅ max_depth: standard / empty")

    sym = TreeNode.from_list([1,2,2,3,4,4,3])
    assert is_symmetric(sym)
    asym = TreeNode.from_list([1,2,2,None,3,None,3])
    assert not is_symmetric(asym)
    print("  ✅ is_symmetric: symmetric / asymmetric")

    d_root = TreeNode.from_list([1,2,3,4,5])
    assert diameter(d_root) == 3
    print("  ✅ diameter")

    lo_root = TreeNode.from_list([3,9,20,None,None,15,7])
    assert level_order(lo_root) == [[3],[9,20],[15,7]]
    assert level_order(None) == []
    print("  ✅ level_order: standard / empty")

    rv = TreeNode.from_list([1,2,3,None,5,None,4])
    assert right_side_view(rv) == [1,3,4]
    print("  ✅ right_side_view")

    lca_root = TreeNode.from_list([3,5,1,6,2,0,8,None,None,7,4])
    p = lca_root.left          # type: ignore  val=5
    q_node = lca_root.right    # type: ignore  val=1
    assert lca(lca_root, p, q_node).val == 3  # type: ignore
    print("  ✅ lca")

    rebuilt = build_from_preorder_inorder([3,9,20,15,7],[9,3,15,20,7])
    assert level_order(rebuilt) == [[3],[9,20],[15,7]]
    print("  ✅ build_from_preorder_inorder")

    kth_root = TreeNode.from_list([3,1,4,None,2])
    assert kth_smallest(kth_root, 1) == 1
    assert kth_smallest(kth_root, 3) == 3
    print("  ✅ kth_smallest")

    mp_root = TreeNode.from_list([-10,9,20,None,None,15,7])
    assert max_path_sum(mp_root) == 42
    assert max_path_sum(TreeNode.from_list([1,2,3])) == 6
    print("  ✅ max_path_sum")

    s_root = TreeNode.from_list([1,2,3,None,None,4,5])
    data = serialize(s_root)
    restored = deserialize(data)
    assert serialize(restored) == data
    print("  ✅ serialize / deserialize")

    assert is_balanced(TreeNode.from_list([3,9,20,None,None,15,7]))
    assert not is_balanced(TreeNode.from_list([1,2,2,3,3,None,None,4,4]))
    print("  ✅ is_balanced: balanced / unbalanced")

    ps_root = TreeNode.from_list([5,4,8,11,None,13,4,7,2,None,None,5,1])
    paths = path_sum(ps_root, 22)
    assert sorted(map(sorted, paths)) == sorted(map(sorted, [[5,4,11,2],[5,8,4,5]]))
    print("  ✅ path_sum")

    print("\n🎉 All tree tests passed!")

if __name__ == "__main__":
    run_tests()

File: trees/trees.ts
/**
 * ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 * TREES  ·  TypeScript
 * ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
 * BST: left < node < right. O(log n) avg, O(n) worst.
 * AVL: self-balancing BST. Guaranteed O(log n).
 * Traversals: inorder(sorted), preorder, postorder, BFS level-order.
 */

class TreeNode {
  val: number; left: TreeNode|null = null; right: TreeNode|null = null;
  constructor(val: number) { this.val = val; }
}

// ── Build helper ──────────────────────────────
function fromArray(vals: (number|null)[]): TreeNode|null {
  if (!vals.length || vals[0] == null) return null;
  const root = new TreeNode(vals[0]);
  const q: TreeNode[] = [root];
  let i = 1;
  while (q.length && i < vals.length) {
    const node = q.shift()!;
    if (i < vals.length && vals[i] != null) { node.left = new TreeNode(vals[i]!); q.push(node.left); }
    i++;
    if (i < vals.length && vals[i] != null) { node.right = new TreeNode(vals[i]!); q.push(node.right); }
    i++;
  }
  return root;
}

// ── BST ───────────────────────────────────────
class BST {
  root: TreeNode|null = null;
  insert(v: number): void { this.root = this._ins(this.root, v); }
  private _ins(n: TreeNode|null, v: number): TreeNode {
    if (!n) return new TreeNode(v);
    if (v < n.val) n.left = this._ins(n.left, v);
    else if (v > n.val) n.right = this._ins(n.right, v);
    return n;
  }
  inorder(): number[] {
    const r: number[] = [];
    const dfs = (n: TreeNode|null) => { if (!n) return; dfs(n.left); r.push(n.val); dfs(n.right); };
    dfs(this.root); return r;
  }
}

// ── Traversals ────────────────────────────────
function inorder(root: TreeNode|null): number[] {
  const r: number[] = [], dfs = (n: TreeNode|null) => { if (!n) return; dfs(n.left); r.push(n.val); dfs(n.right); };
  dfs(root); return r;
}
function levelOrder(root: TreeNode|null): number[][] {
  if (!root) return [];
  const res: number[][] = [], q = [root];
  while (q.length) {
    const level: number[] = [], len = q.length;
    for (let i = 0; i < len; i++) { const n = q.shift()!; level.push(n.val); if (n.left) q.push(n.left); if (n.right) q.push(n.right); }
    res.push(level);
  }
  return res;
}

// ── Problems ──────────────────────────────────
function maxDepth(r: TreeNode|null): number { return r ? 1+Math.max(maxDepth(r.left),maxDepth(r.right)) : 0; }

function invertTree(r: TreeNode|null): TreeNode|null {
  if (!r) return null; [r.left,r.right]=[invertTree(r.right),invertTree(r.left)]; return r;
}

function isSymmetric(r: TreeNode|null): boolean {
  const m = (l: TreeNode|null, ri: TreeNode|null): boolean =>
    !l&&!ri ? true : !l||!ri ? false : l.val===ri.val && m(l.left,ri.right) && m(l.right,ri.left);
  return r ? m(r.left, r.right) : true;
}

function lowestCommonAncestor(r: TreeNode|null, p: TreeNode, q: TreeNode): TreeNode|null {
  if (!r||r===p||r===q) return r;
  const l=lowestCommonAncestor(r.left,p,q), ri=lowestCommonAncestor(r.right,p,q);
  return l&&ri ? r : l??ri;
}

function maxPathSum(r: TreeNode|null): number {
  let best = -Infinity;
  const g = (n: TreeNode|null): number => {
    if (!n) return 0;
    const l=Math.max(g(n.left),0), ri=Math.max(g(n.right),0);
    best=Math.max(best,l+ri+n.val); return n.val+Math.max(l,ri);
  };
  g(r); return best;
}

function rightSideView(r: TreeNode|null): number[] {
  const res: number[] = []; if (!r) return res;
  const q=[r];
  while (q.length) { const len=q.length; for (let i=0;i<len;i++){const n=q.shift()!;if(i===len-1)res.push(n.val);if(n.left)q.push(n.left);if(n.right)q.push(n.right);}}
  return res;
}

function kthSmallest(r: TreeNode|null, k: number): number {
  const stack: TreeNode[] = []; let cur = r, cnt = 0;
  while (stack.length||cur) {
    while (cur) { stack.push(cur); cur=cur.left; }
    cur=stack.pop()!; if(++cnt===k) return cur.val; cur=cur.right;
  }
  return -1;
}

function isValidBST(r: TreeNode|null): boolean {
  const v = (n: TreeNode|null, lo: number, hi: number): boolean =>
    !n ? true : n.val>lo&&n.val<hi && v(n.left,lo,n.val) && v(n.right,n.val,hi);
  return v(r,-Infinity,Infinity);
}

function buildTreeFromPreIn(pre: number[], ino: number[]): TreeNode|null {
  if (!pre.length) return null;
  const root = new TreeNode(pre[0]), mid = ino.indexOf(pre[0]);
  root.left = buildTreeFromPreIn(pre.slice(1,mid+1), ino.slice(0,mid));
  root.right = buildTreeFromPreIn(pre.slice(mid+1), ino.slice(mid+1));
  return root;
}

function diameterOfBinaryTree(r: TreeNode|null): number {
  let best = 0;
  const h = (n: TreeNode|null): number => { if (!n) return 0; const l=h(n.left),ri=h(n.right); best=Math.max(best,l+ri); return 1+Math.max(l,ri); };
  h(r); return best;
}

function isBalanced(r: TreeNode|null): boolean {
  const h = (n: TreeNode|null): number => {
    if (!n) return 0;
    const l=h(n.left), ri=h(n.right);
    if (l<0||ri<0||Math.abs(l-ri)>1) return -1;
    return 1+Math.max(l,ri);
  };
  return h(r) >= 0;
}

function zigzagLevelOrder(r: TreeNode|null): number[][] {
  const res: number[][] = []; if (!r) return res;
  const q=[r]; let leftToRight=true;
  while(q.length){const len=q.length;const level:number[]=[];for(let i=0;i<len;i++){const n=q.shift()!;leftToRight?level.push(n.val):level.unshift(n.val);if(n.left)q.push(n.left);if(n.right)q.push(n.right);}res.push(level);leftToRight=!leftToRight;}
  return res;
}

// ── Tests ─────────────────────────────────────
function assert(c: boolean, m: string) { if (!c) throw new Error(`FAIL: ${m}`); }
const eq = (a: unknown, b: unknown) => JSON.stringify(a)===JSON.stringify(b);

function runTests() {
  console.log("Running tree tests...\n");

  const bst = new BST();
  [5,3,7,1,4,6,8].forEach(v=>bst.insert(v));
  assert(eq(bst.inorder(),[1,3,4,5,6,7,8]),"bst inorder");
  console.log("  ✅ BST: insert / inorder");

  const t = fromArray([3,9,20,null,null,15,7]);
  assert(maxDepth(t)===3,"maxDepth");
  assert(eq(levelOrder(t),[[3],[9,20],[15,7]]),"levelOrder");
  console.log("  ✅ maxDepth / levelOrder");

  const inv = fromArray([4,2,7,1,3,6,9]);
  invertTree(inv);
  assert(inorder(inv)[0]===9,"invertTree");
  console.log("  ✅ invertTree");

  assert(isSymmetric(fromArray([1,2,2,3,4,4,3])),"symmetric");
  assert(!isSymmetric(fromArray([1,2,2,null,3,null,3])),"asymmetric");
  console.log("  ✅ isSymmetric");

  const bst2 = fromArray([6,2,8,0,4,7,9,null,null,3,5]);
  const p=new TreeNode(2),q2=new TreeNode(8);
  const lcaRoot=fromArray([6,2,8,0,4,7,9,null,null,3,5]);
  assert(isValidBST(fromArray([2,1,3])),"validBST");
  assert(!isValidBST(fromArray([5,1,4,null,null,3,6])),"invalidBST");
  console.log("  ✅ isValidBST");

  assert(maxPathSum(fromArray([-10,9,20,null,null,15,7]))===42,"maxPathSum");
  console.log("  ✅ maxPathSum");

  assert(eq(rightSideView(fromArray([1,2,3,null,5,null,4])),[1,3,4]),"rightSideView");
  console.log("  ✅ rightSideView");

  const bst3 = new BST(); [3,1,4,null,2].forEach(v=>v&&bst3.insert(v));
  assert(kthSmallest(bst3.root,1)===1,"kthSmallest");
  console.log("  ✅ kthSmallest");

  assert(diameterOfBinaryTree(fromArray([1,2,3,4,5]))===3,"diameter");
  console.log("  ✅ diameterOfBinaryTree");

  assert(isBalanced(fromArray([3,9,20,null,null,15,7])),"balanced");
  assert(!isBalanced(fromArray([1,2,2,3,3,null,null,4,4])),"unbalanced");
  console.log("  ✅ isBalanced");

  assert(eq(buildTreeFromPreIn([3,9,20,15,7],[9,3,15,20,7]),fromArray([3,9,20,null,null,15,7])),"buildFromPreIn");
  console.log("  ✅ buildFromPreIn");

  assert(eq(zigzagLevelOrder(fromArray([3,9,20,null,null,15,7])),[[3],[20,9],[15,7]]),"zigzag");
  console.log("  ✅ zigzagLevelOrder");

  console.log("\n🎉 All tree tests passed!");
}
runTests();
export { TreeNode, BST, fromArray, inorder, levelOrder, maxDepth, invertTree, isSymmetric, lowestCommonAncestor, maxPathSum, rightSideView, kthSmallest, isValidBST, buildTreeFromPreIn, diameterOfBinaryTree, isBalanced, zigzagLevelOrder };
